cmd/snapd-generator/main.c

```diff
@@ -97,3 +90,80 @@ int main(int argc, char **argv)
 	fprintf(f, "WantedBy=local-fs.target\n");
 	return 0;
 }
+
+int ensure_fusesquashfs_inside_container(const char *normal_dir)
+{
+	// check if we are running inside a container
+	if (system("systemd-detect-virt --container --quiet") != 0) {
+		return 0;
+	}
+
+	char *fstype = NULL;
+	if (system("which squashfuse > /dev/null") == 0) {
```

Comment:
I'm not super fond of using which for this. Could we please not do this via `system`?

Reply:
I was thinking of something like:
```
diff --git a/cmd/snapd-generator/main.c b/cmd/snapd-generator/main.c
index 9e9e7a3714..fd53a4ceb4 100644
--- a/cmd/snapd-generator/main.c
+++ b/cmd/snapd-generator/main.c
@@ -91,17 +91,46 @@ int ensure_root_fs_shared(const char *normal_dir)
        return 0;
 }
 
+static bool file_exists(const char *path) {
+        struct stat buf;
+        return stat(path, &buf) != 0;
+}
+
+static bool in_path(const char *needle) {
+        char *path_copy SC_CLEANUP(sc_cleanup_string) = NULL;
+        path_copy = sc_strdup(getenv("PATH"));
+
+        // strtok_r needs a pointer to keep track of where it is in the string.
+       char *path_walker = NULL;
+
+       // Initialize tokenizer and obtain first path segment.
+       char *path_segment = strtok_r(path_copy, ":", &path_walker);
+       while (path_segment) {
+               char p[PATH_MAX] = { 0 };
+               sc_must_snprintf(p, sizeof(p), "%s/%s",
+                        path_segment, needle);
+               if (file_exists(p)) {
+                      return true;
+               }
+               // Obtain the next path segment.
+               path_segment = strtok_r(NULL, "/", &path_walker);
+        }
+        return false;
+}
+
 int ensure_fusesquashfs_inside_container(const char *normal_dir)
 {
-       // check if we are running inside a container
-       if (system("systemd-detect-virt --container --quiet") != 0) {
+       // check if we are running inside a container, systemd
+        // provides this file all the way back to trusty if run in a
+        // container
+        if (!file_exists("/run/systemd/container")) {
                return 0;
        }
 
        char *fstype = NULL;
-       if (system("which squashfuse > /dev/null") == 0) {
+       if (is_path("squashfuse") == 0) {
                fstype = "fuse.squashfuse";
-       } else if (system("which snapfuse > /dev/null") == 0) {
+       } else if (in_path("snapfuse") == 0) {
                fstype = "fuse.snapfuse";
        } else {
                fprintf(stderr,
```
but of course this is way more code and harder to review :/ I don't know of a simpler way right now (other than system("command -v") or system("which")).

Reply:
Thanks, yes, I implemented something along these lines (slightly different, I didn't notice your comment until having it almost ready).